runtime/near-vm-runner/src/cache.rs

```diff
@@ -275,25 +327,22 @@ pub mod wasmer2_cache {
         config: &VMConfig,
         cache: Option<&dyn CompiledContractCache>,
         store: &wasmer::Store,
-    ) -> Result<wasmer::Module, VMError> {
-        if cache.is_none() {
-            return compile_module_wasmer2(wasm_code, config, store);
-        }
-
-        let cache = cache.unwrap();
-        match cache.get(&key.0) {
-            Ok(serialized) => match serialized {
-                Some(serialized) => deserialize_wasmer2(serialized.as_slice(), store)
-                    .map_err(VMError::CacheError)?,
-                None => compile_and_serialize_wasmer2(wasm_code, &key, config, cache, store),
-            },
-            Err(_) => Err(VMError::CacheError(ReadError)),
+    ) -> Result<Result<wasmer::Module, CompilationError>, CacheError> {
+        match cache {
+            None => Ok(compile_module_wasmer2(wasm_code, config, store)),
+            Some(cache) => {
+                let serialized = cache.get(&key.0).map_err(|_io_err| CacheError::WriteError)?;
+                match serialized {
+                    Some(serialized) => deserialize_wasmer2(serialized.as_slice(), store),
+                    None => compile_and_serialize_wasmer2(wasm_code, &key, config, cache, store),
+                }
+            }
         }
     }
 
     #[cfg(not(feature = "no_cache"))]
     cached_key! {
-        MODULES: SizedCache<CryptoHash, Result<wasmer::Module, VMError>>
+        MODULES: SizedCache<CryptoHash, Result<Result<wasmer::Module, CompilationError>, CacheError>>
```

Comment:
This (both old and new code) I think is wrong -- we shouldn't be storing fatal cache errors in memory cache, we should just always propagate this. This is the reason is why we drop the CompiledContractCache's io::Error from CacheError -- CacheError needs to be clone to store it in in-memory cache. 

Don't want to fix this in this PR, as figuring out how to implement correct caching with this macro is non-trivial.